client/src/pages/files/addTags.tsx

```diff
@@ -0,0 +1,206 @@
+import Prose from 'components/Prose';
+import React, {
+  useState,
+  useEffect,
+  KeyboardEvent,
+  useRef,
+  useMemo,
+} from 'react';
+
+import {getUserFiles, updateDocumentTags} from 'lib/api/files';
+import {UserFile} from 'types/UserFile';
+import {Table} from 'semantic-ui-react';
+
+import {useAuth} from 'contexts/auth';
+
+export default function TaggingView() {
+  const {user} = useAuth();
+  const [userFiles, setUserFiles] = useState<Map<string, UserFile>>(
+    new Map<string, UserFile>()
+  );
+  // Set of all filenames that will change in this update
+  const [filesToChange, setFilesToChange] = useState<Set<string>>(
+    new Set<string>()
+  );
+  useEffect(() => {
+    if (user) {
+      getUserUploadedFiles();
+    }
+  }, [user]);
+
+  async function getUserUploadedFiles() {
+    const newUserFiles = await getUserFiles(user!);
+    setUserFiles(
+      new Map(
+        Array.from(newUserFiles).map(userFile => {
+          return [userFile.fileName, userFile];
+        })
+      )
+    );
+  }
+
+  const removeTagFromFile = (tagIndex: number, fileName: string) => {
+    const newFiles = new Map(userFiles);
+    const oldFileInfo = newFiles.get(fileName)!;
+    const newFileInfo: UserFile = {
+      ...oldFileInfo,
+      tags: oldFileInfo.tags.slice(),
+    };
+    newFileInfo.tags.splice(tagIndex, 1);
+    newFiles.set(oldFileInfo.fileName, newFileInfo);
+    setUserFiles(newFiles);
+
+    const newFilesToChange = new Set<string>(filesToChange);
+    newFilesToChange.add(fileName);
+    setFilesToChange(newFilesToChange);
+  };
+
+  const addTagToFile = (tag: string, fileName: string) => {
+    if (
+      userFiles
+        .get(fileName)!
+        .tags.find(fileTag => fileTag.toLowerCase() === tag.toLowerCase())
+    ) {
+      return false; // False if nothing added
+    }
+    const newFiles = new Map(userFiles);
+    const oldFileInfo = newFiles.get(fileName)!;
+    const newFileInfo: UserFile = {
+      ...oldFileInfo,
+      tags: [...oldFileInfo.tags.slice(), tag],
+    };
+    newFiles.set(oldFileInfo.fileName, newFileInfo);
+    setUserFiles(newFiles);
+
+    const newFilesToChange = new Set<string>(filesToChange);
+    newFilesToChange.add(fileName);
+    setFilesToChange(newFilesToChange);
+
+    return true; // True if a new tag is added
+  };
+
+  const canUploadTags = useMemo(() => !(user === null), [user]);
+
+  const uploadAllTags = async () => {
+    if (!canUploadTags) {
+      return;
+    }
+
+    const allPromises: Promise<void>[] = Array.from(filesToChange).map(
```

Comment:
Nit: I'd call name this variable something describing what it holds, e.g. `uploads`